### PR TITLE
Replace list.sort(cmp=) with list.sort(key=)

### PR DESCRIPTION
**Pull request by rodrigc**
The cmp= keyword argument to list.sort() is gone in Python 3.

See:
http://python3porting.com/problems.html#sorting

### CODE REVIEW

File: master/buildbot/process/buildrequestdistributor.py [hunk 1]
```diff
@@ -404,13 +406,20 @@ def xform(bldr):
...
         # the list
-        def nonecmp(a, b):
+        def nonekey(a):
             if a[0] is None:
-                return 1
-            if b[0] is None:
-                return -1
-            return cmp(a, b)
-        xformed.sort(cmp=nonecmp)
+                b = list(a)
```
Comment (tardyp) [inline master/buildbot/process/buildrequestdistributor.py]: mmh this implementation looks like much more complicated than it was initially with cmp, and much slower.
I would prefer to find an implementation. 
maybe this:
http://code.activestate.com/recipes/576653-convert-a-cmp-function-to-a-key-function/

**Comment by tardyp** [comment]
@rodrigc any update on finishing this?

File: master/buildbot/process/buildrequestdistributor.py (hunk 1 above)
Comment (rodrigc) [inline master/buildbot/process/buildrequestdistributor.py]: I don't think this implementation is very complicated.
If no date is specified in an item, then the sort is supposed to place that item near the end of the list.  This behavior is preserved with my change.

Regarding the Activatestate code example you posted, cmp_to_key() does not exist
in Python 2.6, nor does it exist in the future module, so it cannot be used.

File: master/buildbot/changes/changes.py [hunk 2]
```diff
@@ -162,7 +162,7 @@ def asText(self):
     def asDict(self):
         '''returns a dictonary with suitable info for html/mail rendering'''
         files = [dict(name=f) for f in self.files]
-        files.sort(cmp=lambda a, b: a['name'] < b['name'])
+        files.sort(key=lambda a: a['name'])
```
Comment (sa2ajj) [inline master/buildbot/changes/changes.py]: `operator` module offers help here.

File: master/buildbot/process/buildrequestdistributor.py [hunk 3]
```diff
@@ -13,6 +13,8 @@
 #
 # Copyright Buildbot Team Members
 import random
+from datetime import datetime
```
Comment (sa2ajj) [inline master/buildbot/process/buildrequestdistributor.py]: did `isort` say nothing about these imports?

File: master/buildbot/status/master.py [hunk 4]
```diff
@@ -307,7 +307,7 @@ def refill():
             candidates = [(i, b, b.getTimes()[1])
                           for i, b in enumerate(next_build)
                           if b is not None]
-            candidates.sort(lambda x, y: cmp(x[2], y[2]))
+            candidates.sort(key=lambda x: x[2])
```
Comment (sa2ajj) [inline master/buildbot/status/master.py]: `lambda`s are evil: if you touch them, get rid of them.

File: master/buildbot/test/unit/test_process_buildrequestdistributor.py [hunk 5]
```diff
@@ -35,7 +35,7 @@ def pick_nth_by_name(builder, workers=None, br=None):
         if workers is None:
             workers = builder
         workers = workers[:]
-        workers.sort(cmp=lambda a, b: cmp(a.name, b.name))
+        workers.sort(key=lambda a: a.name)
```
Comment (sa2ajj) [inline master/buildbot/test/unit/test_process_buildrequestdistributor.py]: ditto

File: master/buildbot/test/unit/test_process_buildrequestdistributor.py [hunk 6]
```diff
@@ -210,7 +210,7 @@ def test_maybeStartBuildsOn_parallel(self):
         builders = ['bldr%02d' % i for i in range(15)]
 
         def slow_sorter(master, bldrs):
-            bldrs.sort(lambda b1, b2: cmp(b1.name, b2.name))
+            bldrs.sort(key=lambda b1: b1.name)
```
Comment (sa2ajj) [inline master/buildbot/test/unit/test_process_buildrequestdistributor.py]: ditto

File: master/buildbot/test/unit/test_process_results.py [hunk 7]
```diff
@@ -27,9 +27,15 @@ def test_Results(self):
...
+        self.assertEqual(results.WARNINGS,
+            results.worst_status(results.SUCCESS, results.WARNINGS))
+        self.assertEqual(results.CANCELLED,
+            results.worst_status(results.SKIPPED, results.CANCELLED))
+
+    def test_sort_worst_status(self):
         res = lrange(len(results.Results))
         res.sort(
-            cmp=lambda a, b: 1 if (results.worst_status(a, b) == a) else -1)
+            key=lambda a: a if a != results.SKIPPED else -1)
```
Comment (sa2ajj) [inline master/buildbot/test/unit/test_process_results.py]: ditto

File: master/buildbot/process/buildrequestdistributor.py (hunk 3 above)
Comment (rodrigc) [inline master/buildbot/process/buildrequestdistributor.py]: @sa2ajj isort said nothing about these imports

File: master/buildbot/status/master.py (hunk 4 above)
Comment (rodrigc) [inline master/buildbot/status/master.py]: I don't agree with your comment on lambdas.  Is that your personal opinion, or a position that the buildbot project takes?

Comment (tardyp) [inline master/buildbot/status/master.py]: I agree with @rodrigc with use lambda is lot of place where it makes the code easier to read like here.

**Pull request merged**